File: python/rocketmq/remoting/protocol/subscription.py
# -*- coding: utf-8 -*-
"""订阅组相关模型，对应 org.apache.rocketmq.remoting.protocol.subscription 包：
SubscriptionGroupConfig / GroupRetryPolicy / SimpleSubscriptionData。

字段名与默认值以 Java 5.x 探针实测为准（``JSON.toJSONString(new SubscriptionGroupConfig())``）：
{"attributes":{},"brokerId":0,"consumeBroadcastEnable":true,"consumeEnable":true,
 "consumeFromMinEnable":true,"consumeMessageOrderly":false,"consumeTimeoutMinute":15,
 "groupName":"MyGroup","groupRetryPolicy":{"type":"CUSTOMIZED"},"groupSysFlag":0,
 "notifyConsumerIdsChangedEnable":true,"retryMaxTimes":16,"retryQueueNums":1,
 "whichBrokerWhenConsumeSlowly":1}
注意 fastjson2 **跳过 null 字段**（``subscriptionDataSet`` 为 null 时整个键不出现），
所以 from_dict 一律用 ``.get(k, default)``。
"""
from __future__ import annotations

import json
from typing import Any, Dict, List, Optional

from ...remoting.protocol.serialize import RemotingSerializable

# MixAll.MASTER_ID
MASTER_ID = 0


class GroupRetryPolicyType:
    EXPONENTIAL = "EXPONENTIAL"
    CUSTOMIZED = "CUSTOMIZED"


class GroupRetryPolicy:
    """简化版：只保留 type 与两个子策略的原始 dict（Java 侧为 null 时不序列化）。"""

    def __init__(self, policy_type: str = GroupRetryPolicyType.CUSTOMIZED,
                 exponential_retry_policy: Optional[dict] = None,
                 customized_retry_policy: Optional[dict] = None):
        self.type = policy_type
        self.exponential_retry_policy = exponential_retry_policy
        self.customized_retry_policy = customized_retry_policy

    def to_dict(self) -> dict:
        d = {"type": self.type}
        # Java 默认 type=CUSTOMIZED 且 customizedRetryPolicy 为 null（字段有默认实例但
        # fastjson2 只序列化非 null 的），探针输出里只有 type，故这里同样只在显式赋值时输出。
        if self.exponential_retry_policy is not None:
            d["exponentialRetryPolicy"] = self.exponential_retry_policy
        if self.customized_retry_policy is not None:
            d["customizedRetryPolicy"] = self.customized_retry_policy
        return d

    @staticmethod
    def from_dict(d: Optional[dict]) -> "GroupRetryPolicy":
        d = d or {}
        return GroupRetryPolicy(
            policy_type=d.get("type", GroupRetryPolicyType.CUSTOMIZED),
            exponential_retry_policy=d.get("exponentialRetryPolicy"),
            customized_retry_policy=d.get("customizedRetryPolicy"),
        )


class SimpleSubscriptionData:
    """对应 org.apache.rocketmq.remoting.protocol.subscription.SimpleSubscriptionData。"""

    def __init__(self, topic: str = "", expression_type: str = "TAG",
                 expression: str = "*", version: int = 0):
        self.topic = topic
        self.expression_type = expression_type
        self.expression = expression
        self.version = version

    def to_dict(self) -> dict:
        return {
            "topic": self.topic,
            "expressionType": self.expression_type,
            "expression": self.expression,
            "version": self.version,
        }

    @staticmethod
    def from_dict(d: dict) -> "SimpleSubscriptionData":
        return SimpleSubscriptionData(
            topic=d.get("topic", ""),
            expression_type=d.get("expressionType", "TAG"),
            expression=d.get("expression", "*"),
            version=d.get("version", 0),
        )


class SubscriptionGroupConfig:
    """对应 org.apache.rocketmq.remoting.protocol.subscription.SubscriptionGroupConfig。"""

    def __init__(self, group_name: str = ""):
        self.group_name = group_name
        self.consume_enable = True
        self.consume_from_min_enable = True
        self.consume_broadcast_enable = True
        self.consume_message_orderly = False
        self.retry_queue_nums = 1
        self.retry_max_times = 16
        self.group_retry_policy = GroupRetryPolicy()
        self.broker_id = MASTER_ID
        self.which_broker_when_consume_slowly = 1
        self.notify_consumer_ids_changed_enable = True
        self.group_sys_flag = 0
        self.consume_timeout_minute = 15
        self.subscription_data_set: Optional[List[SimpleSubscriptionData]] = None
        self.attributes: Dict[str, str] = {}

    def to_dict(self) -> dict:
        d = {
            "groupName": self.group_name,
            "consumeEnable": self.consume_enable,
            "consumeFromMinEnable": self.consume_from_min_enable,
            "consumeBroadcastEnable": self.consume_broadcast_enable,
            "consumeMessageOrderly": self.consume_message_orderly,
            "retryQueueNums": self.retry_queue_nums,
            "retryMaxTimes": self.retry_max_times,
            "groupRetryPolicy": self.group_retry_policy.to_dict(),
            "brokerId": self.broker_id,
            "whichBrokerWhenConsumeSlowly": self.which_broker_when_consume_slowly,
            "notifyConsumerIdsChangedEnable": self.notify_consumer_ids_changed_enable,
            "groupSysFlag": self.group_sys_flag,
            "consumeTimeoutMinute": self.consume_timeout_minute,
            "attributes": self.attributes,
        }
        # fastjson2 默认跳过 null
        if self.subscription_data_set is not None:
            d["subscriptionDataSet"] = [s.to_dict() for s in self.subscription_data_set]
        return d

    @staticmethod
    def from_dict(d: dict) -> "SubscriptionGroupConfig":
        cfg = SubscriptionGroupConfig(d.get("groupName", ""))
        cfg.consume_enable = d.get("consumeEnable", True)
        cfg.consume_from_min_enable = d.get("consumeFromMinEnable", True)
        cfg.consume_broadcast_enable = d.get("consumeBroadcastEnable", True)
        cfg.consume_message_orderly = d.get("consumeMessageOrderly", False)
        cfg.retry_queue_nums = d.get("retryQueueNums", 1)
        cfg.retry_max_times = d.get("retryMaxTimes", 16)
        cfg.group_retry_policy = GroupRetryPolicy.from_dict(d.get("groupRetryPolicy"))
        cfg.broker_id = d.get("brokerId", MASTER_ID)
        cfg.which_broker_when_consume_slowly = d.get("whichBrokerWhenConsumeSlowly", 1)
        cfg.notify_consumer_ids_changed_enable = d.get("notifyConsumerIdsChangedEnable", True)
        cfg.group_sys_flag = d.get("groupSysFlag", 0)
        cfg.consume_timeout_minute = d.get("consumeTimeoutMinute", 15)
        raw_sub = d.get("subscriptionDataSet")
        cfg.subscription_data_set = (
            [SimpleSubscriptionData.from_dict(x) for x in raw_sub] if raw_sub else None)
        cfg.attributes = dict(d.get("attributes") or {})
        return cfg

    def encode(self) -> bytes:
        return RemotingSerializable.encode(self.to_dict())

    @staticmethod
    def decode(data: bytes) -> "SubscriptionGroupConfig":
        return SubscriptionGroupConfig.from_dict(RemotingSerializable.decode_json(data))

    def __repr__(self):
        return "SubscriptionGroupConfig[groupName=%s, retryQueueNums=%d, brokerId=%d]" % (
            self.group_name, self.retry_queue_nums, self.broker_id)


class SubscriptionGroupWrapper:
    """对应 org.apache.rocketmq.remoting.protocol.body.SubscriptionGroupWrapper。

    探针输出：{"dataVersion":{...},"forbiddenTable":{},"subscriptionGroupTable":{...}}
    """

    def __init__(self):
        self.subscription_group_table: Dict[str, SubscriptionGroupConfig] = {}
        self.forbidden_table: Dict[str, Any] = {}
        self.data_version: Dict[str, Any] = {}

    def to_dict(self) -> dict:
        return {
            "dataVersion": self.data_version,
            "forbiddenTable": self.forbidden_table,
            "subscriptionGroupTable": {k: v.to_dict() for k, v in self.subscription_group_table.items()},
        }

    @staticmethod
    def from_dict(d: dict) -> "SubscriptionGroupWrapper":
        w = SubscriptionGroupWrapper()
        table = d.get("subscriptionGroupTable") or {}
        w.subscription_group_table = {
            k: SubscriptionGroupConfig.from_dict(v) for k, v in table.items()}
        w.forbidden_table = dict(d.get("forbiddenTable") or {})
        w.data_version = dict(d.get("dataVersion") or {})
        return w

    def encode(self) -> bytes:
        return RemotingSerializable.encode(self.to_dict())

    @staticmethod
    def decode(data: bytes) -> "SubscriptionGroupWrapper":
        return SubscriptionGroupWrapper.from_dict(RemotingSerializable.decode_json(data))


__all__ = ["SubscriptionGroupConfig", "SubscriptionGroupWrapper",
           "GroupRetryPolicy", "GroupRetryPolicyType", "SimpleSubscriptionData",
           "MASTER_ID"]

File: dotnet/src/RocketMQ.Client/Remoting/Protocol/Subscription.cs
// 订阅组模型（对应 org.apache.rocketmq.remoting.protocol.subscription 包）：
// SubscriptionGroupConfig / GroupRetryPolicy / SimpleSubscriptionData /
// SubscriptionGroupWrapper。
//
// 字段名与默认值以 Java 5.x 探针实测为准：
//   JSON.toJSONString(new SubscriptionGroupConfig("MyGroup")) ->
//   {"attributes":{},"brokerId":0,"consumeBroadcastEnable":true,"consumeEnable":true,
//    "consumeFromMinEnable":true,"consumeMessageOrderly":false,"consumeTimeoutMinute":15,
//    "groupName":"MyGroup","groupRetryPolicy":{"type":"CUSTOMIZED"},"groupSysFlag":0,
//    "notifyConsumerIdsChangedEnable":true,"retryMaxTimes":16,"retryQueueNums":1,
//    "whichBrokerWhenConsumeSlowly":1}
// fastjson2 **跳过 null 字段**（subscriptionDataSet 为 null 时整个键不出现）。
using System.Globalization;
using RocketMQ.Common;

namespace RocketMQ.Remoting.Protocol;

/// <summary>org.apache.rocketmq.remoting.protocol.subscription.GroupRetryPolicyType。</summary>
public static class GroupRetryPolicyType
{
    public const string Exponential = "EXPONENTIAL";
    public const string Customized = "CUSTOMIZED";
}

/// <summary>
/// 对应 GroupRetryPolicy。两个子策略在 Java 里默认是 new 出来的实例，但 fastjson2
/// 的探针输出里只有 {"type":"CUSTOMIZED"} —— 说明它们实际序列化时被跳过。
/// 因此这里同样只在显式设置时输出（用 JsonValue 透传，避免过度建模）。
/// </summary>
public sealed class GroupRetryPolicy
{
    public string Type { get; set; } = GroupRetryPolicyType.Customized;

    // IsNull 表示不输出
    public JsonValue ExponentialRetryPolicy { get; set; } = JsonValue.Null;

    // IsNull 表示不输出
    public JsonValue CustomizedRetryPolicy { get; set; } = JsonValue.Null;

    public JsonValue ToJson()
    {
        var v = JsonValue.MakeObject();
        v.Set("type", JsonValue.MakeString(Type));
        if (!ExponentialRetryPolicy.IsNull)
        {
            v.Set("exponentialRetryPolicy", ExponentialRetryPolicy);
        }

        if (!CustomizedRetryPolicy.IsNull)
        {
            v.Set("customizedRetryPolicy", CustomizedRetryPolicy);
        }

        return v;
    }

    public static GroupRetryPolicy FromJson(JsonValue v)
    {
        var p = new GroupRetryPolicy();
        if (!v.IsObject)
        {
            return p;
        }

        p.Type = v.TryGetString("type", out string t) ? t : GroupRetryPolicyType.Customized;
        p.ExponentialRetryPolicy = v.Get("exponentialRetryPolicy");
        p.CustomizedRetryPolicy = v.Get("customizedRetryPolicy");
        return p;
    }
}

/// <summary>对应 SimpleSubscriptionData。</summary>
public sealed class SimpleSubscriptionData
{
    public string Topic { get; set; } = string.Empty;
    public string ExpressionType { get; set; } = "TAG";
    public string Expression { get; set; } = "*";
    public long Version { get; set; }

    public JsonValue ToJson()
    {
        var v = JsonValue.MakeObject();
        v.Set("topic", JsonValue.MakeString(Topic));
        v.Set("expressionType", JsonValue.MakeString(ExpressionType));
        v.Set("expression", JsonValue.MakeString(Expression));
        v.Set("version", JsonValue.MakeInt(Version));
        return v;
    }

    public static SimpleSubscriptionData FromJson(JsonValue v) =>
        new()
        {
            Topic = v.Get("topic").StringValue(),
            ExpressionType = v.Get("expressionType").StringValue("TAG"),
            Expression = v.Get("expression").StringValue("*"),
            Version = v.Get("version").IntValue(0),
        };
}

/// <summary>对应 SubscriptionGroupConfig。</summary>
public sealed class SubscriptionGroupConfig
{
    // MixAll.MASTER_ID
    public const int MasterId = 0;

    public string GroupName { get; set; } = string.Empty;
    public bool ConsumeEnable { get; set; } = true;
    public bool ConsumeFromMinEnable { get; set; } = true;
    public bool ConsumeBroadcastEnable { get; set; } = true;
    public bool ConsumeMessageOrderly { get; set; }
    public int RetryQueueNums { get; set; } = 1;
    public int RetryMaxTimes { get; set; } = 16;
    public GroupRetryPolicy GroupRetryPolicy { get; set; } = new();
    public int BrokerId { get; set; } = MasterId;
    public int WhichBrokerWhenConsumeSlowly { get; set; } = 1;
    public bool NotifyConsumerIdsChangedEnable { get; set; } = true;
    public int GroupSysFlag { get; set; }
    public int ConsumeTimeoutMinute { get; set; } = 15;
    public PropertyMap Attributes { get; set; } = new();

    // Java 为 null 时 fastjson2 整个键不输出
    public List<SimpleSubscriptionData> SubscriptionDataSet { get; } = new();
    public bool HasSubscriptionDataSet { get; set; }

    public SubscriptionGroupConfig()
    {
    }

    public SubscriptionGroupConfig(string name) => GroupName = name;

    public JsonValue ToJson()
    {
        var v = JsonValue.MakeObject();
        v.Set("groupName", JsonValue.MakeString(GroupName));
        v.Set("consumeEnable", JsonValue.MakeBool(ConsumeEnable));
        v.Set("consumeFromMinEnable", JsonValue.MakeBool(ConsumeFromMinEnable));
        v.Set("consumeBroadcastEnable", JsonValue.MakeBool(ConsumeBroadcastEnable));
        v.Set("consumeMessageOrderly", JsonValue.MakeBool(ConsumeMessageOrderly));
        v.Set("retryQueueNums", JsonValue.MakeInt(RetryQueueNums));
        v.Set("retryMaxTimes", JsonValue.MakeInt(RetryMaxTimes));
        v.Set("groupRetryPolicy", GroupRetryPolicy.ToJson());
        v.Set("brokerId", JsonValue.MakeInt(BrokerId));
        v.Set("whichBrokerWhenConsumeSlowly", JsonValue.MakeInt(WhichBrokerWhenConsumeSlowly));
        v.Set("notifyConsumerIdsChangedEnable", JsonValue.MakeBool(NotifyConsumerIdsChangedEnable));
        v.Set("groupSysFlag", JsonValue.MakeInt(GroupSysFlag));
        v.Set("consumeTimeoutMinute", JsonValue.MakeInt(ConsumeTimeoutMinute));
        v.Set("attributes", AdminJsonStrings.MakeStringMap(Attributes));
        // fastjson2 默认跳过 null：只有显式设置过才输出该键
        if (HasSubscriptionDataSet)
        {
            var arr = JsonValue.MakeArray();
            foreach (SimpleSubscriptionData s in SubscriptionDataSet)
            {
                arr.PushArray(s.ToJson());
            }

            v.Set("subscriptionDataSet", arr);
        }

        return v;
    }

    public static SubscriptionGroupConfig FromJson(JsonValue v)
    {
        var c = new SubscriptionGroupConfig
        {
            GroupName = v.Get("groupName").StringValue(),
            ConsumeEnable = v.Get("consumeEnable").BoolValue(true),
            ConsumeFromMinEnable = v.Get("consumeFromMinEnable").BoolValue(true),
            ConsumeBroadcastEnable = v.Get("consumeBroadcastEnable").BoolValue(true),
            ConsumeMessageOrderly = v.Get("consumeMessageOrderly").BoolValue(false),
            RetryQueueNums = JavaNumber.ToInt32(v.Get("retryQueueNums").IntValue(1)),
            RetryMaxTimes = JavaNumber.ToInt32(v.Get("retryMaxTimes").IntValue(16)),
            BrokerId = JavaNumber.ToInt32(v.Get("brokerId").IntValue(MasterId)),
            WhichBrokerWhenConsumeSlowly = JavaNumber.ToInt32(v.Get("whichBrokerWhenConsumeSlowly").IntValue(1)),
            NotifyConsumerIdsChangedEnable = v.Get("notifyConsumerIdsChangedEnable").BoolValue(true),
            GroupSysFlag = JavaNumber.ToInt32(v.Get("groupSysFlag").IntValue(0)),
            ConsumeTimeoutMinute = JavaNumber.ToInt32(v.Get("consumeTimeoutMinute").IntValue(15)),
            Attributes = ReadStringMap(v.Get("attributes")),
        };
        JsonValue? policy = v.Find("groupRetryPolicy");
        if (policy is not null)
        {
            c.GroupRetryPolicy = GroupRetryPolicy.FromJson(policy);
        }

        JsonValue? sub = v.Find("subscriptionDataSet");
        if (sub is { IsArray: true } && sub.Size() > 0)
        {
            c.HasSubscriptionDataSet = true;
            c.SubscriptionDataSet.Clear();
            for (int i = 0; i < sub.Size(); ++i)
            {
                c.SubscriptionDataSet.Add(SimpleSubscriptionData.FromJson(sub.At(i)));
            }
        }

        return c;
    }

    private static PropertyMap ReadStringMap(JsonValue v)
    {
        var @out = new PropertyMap();
        JsonValue? p = v.Find("attributes");
        if (p is { IsObject: true })
        {
            foreach (var kv in p.ObjectItems())
            {
                if (kv.Value.IsString)
                {
                    @out[kv.Key] = kv.Value.StringValue();
                }
            }
        }

        return @out;
    }

    public byte[] Encode() => RemotingSerializable.Encode(ToJson());

    public static bool Decode(byte[] data, out SubscriptionGroupConfig @out)
    {
        @out = new SubscriptionGroupConfig();
        if (!RemotingSerializable.Decode(data, out JsonValue v))
        {
            return false;
        }

        @out = FromJson(v);
        return true;
    }

    public override string ToString() =>
        "SubscriptionGroupConfig [groupName=" + GroupName
        + ", consumeEnable=" + (ConsumeEnable ? "true" : "false")
        + ", consumeFromMinEnable=" + (ConsumeFromMinEnable ? "true" : "false")
        + ", consumeBroadcastEnable=" + (ConsumeBroadcastEnable ? "true" : "false")
        + ", consumeMessageOrderly=" + (ConsumeMessageOrderly ? "true" : "false")
        + ", retryQueueNums=" + RetryQueueNums.ToString(CultureInfo.InvariantCulture)
        + ", retryMaxTimes=" + RetryMaxTimes.ToString(CultureInfo.InvariantCulture)
        + ", brokerId=" + BrokerId.ToString(CultureInfo.InvariantCulture)
        + ", whichBrokerWhenConsumeSlowly=" + WhichBrokerWhenConsumeSlowly.ToString(CultureInfo.InvariantCulture) + "]";
}

/// <summary>
/// 对应 org.apache.rocketmq.remoting.protocol.body.SubscriptionGroupWrapper。
/// 探针输出：{"dataVersion":{...},"forbiddenTable":{},"subscriptionGroupTable":{...}}
/// </summary>
public sealed class SubscriptionGroupWrapper
{
    public SortedDictionary<string, SubscriptionGroupConfig> SubscriptionGroupTable { get; } = new();

    // 透传（本实现不解释其内容）
    public JsonValue ForbiddenTable { get; set; } = JsonValue.Null;

    // 透传：翻页请求要原样回传，避免重排/丢精度
    public JsonValue DataVersion { get; set; } = JsonValue.Null;

    public JsonValue ToJson()
    {
        var v = JsonValue.MakeObject();
        v.Set("dataVersion", DataVersion.IsNull ? JsonValue.MakeObject() : DataVersion);
        v.Set("forbiddenTable", ForbiddenTable.IsNull ? JsonValue.MakeObject() : ForbiddenTable);
        var table = JsonValue.MakeObject();
        foreach (var kv in SubscriptionGroupTable)
        {
            table.Set(kv.Key, kv.Value.ToJson());
        }

        v.Set("subscriptionGroupTable", table);
        return v;
    }

    public static SubscriptionGroupWrapper FromJson(JsonValue v)
    {
        var w = new SubscriptionGroupWrapper();
        JsonValue? table = v.Find("subscriptionGroupTable");
        if (table is { IsObject: true })
        {
            foreach (var kv in table.ObjectItems())
            {
                w.SubscriptionGroupTable[kv.Key] = SubscriptionGroupConfig.FromJson(kv.Value);
            }
        }

        w.ForbiddenTable = v.Get("forbiddenTable");
        w.DataVersion = v.Get("dataVersion");
        return w;
    }

    public byte[] Encode() => RemotingSerializable.Encode(ToJson());

    public static bool Decode(byte[] data, out SubscriptionGroupWrapper @out)
    {
        @out = new SubscriptionGroupWrapper();
        if (!RemotingSerializable.Decode(data, out JsonValue v))
        {
            return false;
        }

        @out = FromJson(v);
        return true;
    }

    /// <summary>合并另一页（getAllSubscriptionGroup 分页累积用）。</summary>
    public void MergeFrom(SubscriptionGroupWrapper other)
    {
        foreach (var kv in other.SubscriptionGroupTable)
        {
            SubscriptionGroupTable[kv.Key] = kv.Value;
        }

        if (other.ForbiddenTable.IsObject)
        {
            if (!ForbiddenTable.IsObject)
            {
                ForbiddenTable = JsonValue.MakeObject();
            }

            foreach (var kv in other.ForbiddenTable.ObjectItems())
            {
                ForbiddenTable.Set(kv.Key, kv.Value);
            }
        }

        if (!other.DataVersion.IsNull)
        {
            DataVersion = other.DataVersion;
        }
    }
}
